### DISCUSSION TITLE
Support for Opc UA 1.04

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1242] [General]
Hello, we need in our Milo-Server support for Opc UA Version 1.04 so i would like to ask if it makes sense for us to switch to the Dev1.0 Branch? Because i read it implements 1.04.  So that we later have lesser migration steps when it comes to Milo 1.0 Release or if it makes more sense to wait for the release? 
Thanks in Advance Johann

**Comment by kevinherron** [comment]
What JDK version is your project using?

**Comment by johannwesely** [comment]
We use JDK 17.

**Reply by kevinherron** [reply]
Ok, then it might be worth exploring the dev/1.0 branch. I’m not sure yet how many more API breaking changes are coming, but if you have your own internal artifact repository where you can host a “stable” (i.e. non-SNAPSHOT) artifact to develop against, it might be ok. Of course if you can just wait that will be safer, but more work later.

**Reply by kevinherron** [reply]
I just started work on what I hope is the last big API change in the server. The `AddressSpace` methods implemented by user code are changing from non-blocking to blocking.

e.g.
```java
void read(ReadContext context, Double maxAge, TimestampsToReturn timestamps, List<ReadValueId> readValueIds);
```
becomes
```java
List<DataValue> read(ReadContext context, Double maxAge, TimestampsToReturn timestamps, List<ReadValueId> readValueIds);
```

In practice, so far, it has been a fairly easy change to accommodate in my implementations.
